Source/Core/Core/HW/GCMemcard/GCMemcard.cpp

```diff
@@ -27,196 +27,256 @@ static void ByteSwap(u8* valueA, u8* valueB)
   *valueB = tmp;
 }
 
-GCMemcard::GCMemcard(const std::string& filename, bool forceCreation, bool shift_jis)
-    : m_valid(false), m_filename(filename)
+static constexpr u64 MegabitsToBytes(u64 megabits)
 {
-  // Currently there is a string freeze. instead of adding a new message about needing r/w
-  // open file read only, if write is denied the error will be reported at that point
-  File::IOFile mcdFile(m_filename, "rb");
-  if (!mcdFile.IsOpen())
-  {
-    if (!forceCreation)
-    {
-      if (!AskYesNoT("\"%s\" does not exist.\n Create a new 16MB Memory Card?", filename.c_str()))
-        return;
-      shift_jis =
-          AskYesNoT("Format as Shift JIS (Japanese)?\nChoose no for Windows-1252 (Western)");
-    }
-    Format(shift_jis);
-    return;
-  }
-  else
-  {
-    // This function can be removed once more about hdr is known and we can check for a valid header
-    std::string fileType;
-    SplitPath(filename, nullptr, nullptr, &fileType);
-    if (strcasecmp(fileType.c_str(), ".raw") && strcasecmp(fileType.c_str(), ".gcp"))
-    {
-      PanicAlertT("File has the extension \"%s\".\nValid extensions are (.raw/.gcp)",
-                  fileType.c_str());
-      return;
-    }
-    auto size = mcdFile.GetSize();
-    if (size < MC_FST_BLOCKS * BLOCK_SIZE)
-    {
-      PanicAlertT("%s failed to load as a memory card.\nFile is not large enough to be a valid "
-                  "memory card file (0x%x bytes)",
-                  filename.c_str(), (unsigned)size);
-      return;
-    }
-    if (size % BLOCK_SIZE)
-    {
-      PanicAlertT("%s failed to load as a memory card.\nCard file size is invalid (0x%x bytes)",
-                  filename.c_str(), (unsigned)size);
-      return;
-    }
+  return megabits * ((1024 * 1024) / 8);
+}
 
-    m_size_mb = (u16)((size / BLOCK_SIZE) / MBIT_TO_BLOCKS);
-    switch (m_size_mb)
-    {
-    case MemCard59Mb:
-    case MemCard123Mb:
-    case MemCard251Mb:
-    case Memcard507Mb:
-    case MemCard1019Mb:
-    case MemCard2043Mb:
-      break;
-    default:
-      PanicAlertT("%s failed to load as a memory card.\nCard size is invalid (0x%x bytes)",
-                  filename.c_str(), (unsigned)size);
-      return;
-    }
-  }
+bool GCMemcardErrorCode::HasCriticalErrors() const
+{
+  return Test(GCMemcardValidityIssues::FAILED_TO_OPEN) || Test(GCMemcardValidityIssues::IO_ERROR) ||
```

Comment:
I wonder if this is nicer as (warning, half pseudo-code):
```c++
std::array<GCMemcardValidityIssues, 7> critical_errors {{ FAILED_TO_OPEN, IO_ERROR, ... }};
return std::any_of(critical_errors, Test);
```

I briefly considered suggesting checking whether _only_ non-critical errors are present (since theres only one of those, vs. 7 critical ones); but for clarity this seems the nicer (albeit longer) way.

It's also sad that `std::bitset` has no real mass-test (whether any given number of bits are set), only those that test for all/any/none and specific bits. There is bitwise operators, but I don't think this should be rewritten as `(m_errors & critical_error_flags) != 0` (again for clariry/readabilty)

Reply:
I tried this and ended up with:

~~~
  std::array<GCMemcardValidityIssues, 7> critical_errors{
      {GCMemcardValidityIssues::FAILED_TO_OPEN, GCMemcardValidityIssues::IO_ERROR,
       GCMemcardValidityIssues::INVALID_CARD_SIZE, GCMemcardValidityIssues::INVALID_CHECKSUM,
       GCMemcardValidityIssues::MISMATCHED_CARD_SIZE, GCMemcardValidityIssues::FREE_BLOCK_MISMATCH,
       GCMemcardValidityIssues::DIR_BAT_INCONSISTENT}};
  return std::any_of(critical_errors.begin(), critical_errors.end(),
                     [&](GCMemcardValidityIssues e) { return Test(e); });
~~~

I'm not sure if that's really any better, to be honest.

Reply:
IIRC a trailing comma after the last one lets lint be more lenient and allow one per line.
But I guess I'm ok with either solution at this point, since all of them look kinda ugly in the end.

Reply:
I'll just leave it as-is then.